### PR TITLE
End of Form Error Handling

### PR DESCRIPTION
**Pull request by Jtang-1**
## Product Description

This PR does two things: 
- Displays errors while rebuilding session from frame that were previously caught but not displayed. Namely, errors resulting from end of form navigation but that is not the only feature where rebuilding session is done.
- Adds error handling for when no matches are found between steps and menu options.
Screen Shot 2024-03-21 at 10 26 19 PM

## Technical Summary

[Jira Ticket](https://dimagi.atlassian.net/browse/USH-4176)

### Safety story

tested locally, will test on staging

### Automated test coverage

Basic end of form navigation tests exists and the response is as expected

### QA Plan

no QA

### Special deploy instructions

- [X] This PR can be deployed after merge with no further considerations.

### Rollback instructions

- [X] This PR can be reverted after deploy with no further considerations.

### Review

- [X] The set of people pinged as reviewers is appropriate for the level of risk of the change.

### CODE REVIEW

**Review by snopoke — APPROVED** [review]

**Comment by Jtang-1** [comment]
@snopoke There's an edge case causing test failures and I can't figure out how to best handle it. Could I get your thoughts? From what I understand, the situation is:
Steps `[m1, case_id, case_id_new_child_case_0]`. We just processed `case_id` (via EntityScreen) and move to the next screen which is a MenuScreen['Case Claim' with choices [Entry with id m1-f0, Menu with id m2]].

`case_id_new_child_case_0` is a datum for a new case id (`<datum id="case_id_new_child_case_0" function="uuid()"/>`). The step is `(CASE_ID case_id_new_child_case_0 : 81e0aa64-7aea-4134-b000-43f04fd80f2d)` (where the case_id is a generated uuid). The logic before this PR is that this step is ignored (if this [loop](https://github.com/dimagi/formplayer/blob/767f3becad2563e9ab181fe06389fd819365f223/src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java#L101-L109) has no matches, `currentStep` stays `null` and [we break out of the `while` loop.](https://github.com/dimagi/formplayer/blob/767f3becad2563e9ab181fe06389fd819365f223/src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java#L190-L191))

But with this PR, the non-match throws an exception. A solution for this is: if we encounter a match "miss", we check if the remaining steps are datums for new case ids. If true, we don't throw an exception. However, I don't see a way to know if the step is associated with creating a new case_id. 

But similarly, if the last step is `(SELECTED_ENTITIES selected_cases : ...)`, it causes the same issue. So maybe an additional condition before throwing the error should be that the step we're processing is a `COMMAND_ID`.

**Comment by snopoke** [comment]
I did some debugging and it took a while to figure out what was going on but I think your suggestion to check that the current step we're processing is a `COMMAND_ID` makes sense.

This might be a challenge to implement without completely refactoring that method since we don't actually track which step is the current one. Personally I've wanted to refactor that function many times but never managed to push it all the way through.

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java [hunk 1]
```diff
@@ -106,6 +107,17 @@ public void rebuildSessionFromFrame(MenuSession menuSession, CaseSearchHelper ca
                         }
                     }
                 }
+                if (currentStep == null && processedStepsCount != steps.size()) {
+                    StringJoiner optionsIDJoiner = new StringJoiner(", ", "[", "]");
```
Comment (MartinRiese) [inline src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:111]: I cannot believe `StringJoiner` exists since 1.8 and I never heard of it. Pretty handy.

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java [hunk 2]
```diff
@@ -106,6 +107,17 @@ public void rebuildSessionFromFrame(MenuSession menuSession, CaseSearchHelper ca
                         }
                     }
                 }
+                if (currentStep == null && processedStepsCount != steps.size()) {
```
Comment (MartinRiese) [inline src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:110]: Could you push that in its own method? `rebuildSessionFromFrame` is quite long already

Comment (Jtang-1) [reply src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:115]: [bf6e37e](https://github.com/dimagi/formplayer/pull/1560/commits/bf6e37e2230b0c66ff3b542a85d89a6219b4242b)

**Comment by Jtang-1** [comment]
Thank you for taking the time to look into this for me @snopoke. I ended up tracking what steps were processed, and checking that the remaining unprocessed steps contained a `COMMAND_ID` type (minus the special value `claim_command...`) that was not matched to an available option. This seems to be more conservative to reduce chances of falsely throwing the exception.
From testing, I found tracking the step we're trying to process is tricky (as you mentioned). It seems like there's steps that we just skip and don't handle within `rebuildSessionFromFrame`, such as the `claim_command...` and the "new case id` datums. Unfortunately, I'll also be passing this along and not refactor it at this moment.

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java [hunk 3]
```diff
@@ -101,11 +106,15 @@ public void rebuildSessionFromFrame(MenuSession menuSession, CaseSearchHelper ca
                     for (StackFrameStep step : steps) {
                         if (step.getId().equals(options[i].getCommandID())) {
                             currentStep = String.valueOf(i);
+                            processedSteps.add(step);
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:109]: I notice this loop has no break statement in it. I think it should since I can't think of a scenario where more than one step would match.

Comment (Jtang-1) [reply src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:109]: I was thinking that too. I will make that change in this separate [PR](https://github.com/dimagi/formplayer/pull/1562)

Aside note: I was speaking to Shubham about commit [4ef0364](https://github.com/dimagi/formplayer/pull/1560/commits/4ef0364b1a911e25fd90f45a6a4615e3d15f7d96) and he suggested doing this in a separate PR with more scrutiny (doing QA). So may as well have this extra check done on the `break` statement too.

**Comment by Jtang-1** [comment]
Bumping for review

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java (hunk 3 above)
Comment (snopoke) [reply src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:111]: In that case, should this line be moved out of the for loop since only the last matching step would be 'processed'. I don't think it should be possible for multiple steps to match but I can't prove that.

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java [hunk 4]
```diff
@@ -114,7 +125,19 @@ public void rebuildSessionFromFrame(MenuSession menuSession, CaseSearchHelper ca
                     if (step.getId().equals(neededDatum.getDataId())) {
                         if (entityScreen.referencesContainStep(step.getValue())) {
                             currentStep = step.getValue();
+                            processedSteps.add(step);
                             needsFullInit = ++processedStepsCount == steps.size();
+                        } else {
+                            // This block constructs the message to display then throws the exception
+                            List<String> refsList = entityScreen.getReferences().stream()
+                            .map(ref -> EntityScreen.getReturnValueFromSelection(ref, (EntityDatum) neededDatum, entityScreen.getEvalContext()))
+                            .collect(Collectors.toList());
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:134]: instead of `.collect(Collectors.toList())` you should be able to just call `.toList()`

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java [hunk 5]
```diff
@@ -114,7 +125,19 @@ public void rebuildSessionFromFrame(MenuSession menuSession, CaseSearchHelper ca
                     if (step.getId().equals(neededDatum.getDataId())) {
                         if (entityScreen.referencesContainStep(step.getValue())) {
                             currentStep = step.getValue();
+                            processedSteps.add(step);
                             needsFullInit = ++processedStepsCount == steps.size();
+                        } else {
+                            // This block constructs the message to display then throws the exception
```
Comment (snopoke) [inline src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java:131]: similar to the comment above, it would be good to extract this to a method

**Review by snopoke — APPROVED** [review]
Nothing blocking.

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java (hunk 5 above)
Comment (Jtang-1) [reply src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java]: [6754f61](https://github.com/dimagi/formplayer/pull/1560/commits/6754f61adb48a77d3581fcce4be805c8385918e6)

File: src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java (hunk 4 above)
Comment (Jtang-1) [reply src/main/java/org/commcare/formplayer/services/MenuSessionFactory.java]: [8b47055](https://github.com/dimagi/formplayer/pull/1560/commits/8b47055ee35d51dd9e5d5aad89eb30a9c9c60a81)

**Comment by Jtang-1** [comment]
@snopoke, I couldn't reply to that comment thread
> In that case, should this line be moved out of the for loop since only the last matching step would be 'processed'. I don't think it should be possible for multiple steps to match but I can't prove that.

I'm not entirely sure if it's possible either but I see `needsFullInit = ++processedStepsCount == steps.size();` is within the for loop so I kept it within the same level to be on the safe side.

**Review by snopoke — APPROVED** [review]
